### PR TITLE
VVVVVV: Fix generation reliability issues and increase `DoorCost` max to 5.

### PR DESCRIPTION
**Pull request by ThePhar**
@N00byKing for your review.

## What is this fixing or adding?
Refactor of rules and regions generation code to fix generation reliability failures, this also fixed DoorCost 4 and 5, so re-enabled those settings as client supports them without issue.

Also conditionally marks Trinkets not needed for doors as filler.

## How was this tested?
Generated several all-random 1000 slot V6 seeds and along side some Z3 and RL yamls. No generation failures.

## If this makes graphical changes, please attach screenshots.
image
image
[AP_21993286179638124014_Spoiler.txt](https://github.com/ArchipelagoMW/Archipelago/files/10172339/AP_21993286179638124014_Spoiler.txt)

### CODE REVIEW

**Comment by ThePhar** [comment]
Apparently, I forgot to run unit tests.

File: worlds/v6/Rules.py [hunk 1]
```diff
@@ -1,43 +1,44 @@
...
-        world.random.shuffle(areashuffle)
-    area_cost_map.update({(index + 1): (value + 1) for index, value in enumerate(areashuffle)})
-    area_cost_map.update({0: 0})
-
-    for i in range(1, 5):
-        connect_regions(world, player, "Menu", v6areas[area_connections[i] - 1],
-                        lambda state, i=i: _has_trinket_range(state, player,
-                                                              world.DoorCost[player].value * (area_cost_map[i] - 1),
-                                                              world.DoorCost[player].value * area_cost_map[i]))
+        return state.has(f"Trinket {str(i + 1).zfill(2)}", player)
```
Comment (N00byKing) [inline worlds/v6/Rules.py:8]: Doesn't returning immediately ignore the loop, and could lead to generating impossible seeds?

File: worlds/v6/Rules.py [hunk 2]
```diff
@@ -1,43 +1,44 @@
...
+def can_reach_npc_trinket(state: CollectionState, player: int) -> bool:
+    return state.can_reach("Laboratory", "Entrance", player) or (
+           state.can_reach("The Tower", "Entrance", player) and
+           state.can_reach("Space Station 2", "Entrance", player) and
+           state.can_reach("Warp Zone", "Entrance", player))
+
+
+def set_rules(multiworld: MultiWorld, player: int):
+    for index, region in enumerate(v6_areas, start=1):
+        multiworld.get_entrance(region, player).access_rule = lambda state: has_trinket_range(
```
Comment (N00byKing) [inline worlds/v6/Rules.py:27]: I don't like using `access_rule` directly, please use `add_rule` instead

File: worlds/v6/__init__.py [hunk 3]
```diff
@@ -23,46 +23,62 @@ class V6Web(WebWorld):
...
+    def create_item(self, name: str, progression: bool = False) -> V6Item:
+        return V6Item(
+            name,
+            ItemClassification.progression if progression else ItemClassification.filler,
+            item_table[name],
+            self.player)
+
+    def generate_early(self):
+        self.area_connections = {0: 0, 1: 1, 2: 2, 3: 3, 4: 4}
+        self.area_cost_map = {0: 0, 1: 1, 2: 2, 3: 3, 4: 4}
```
Comment (N00byKing) [inline worlds/v6/__init__.py:56]: `{i : i for i in range(5)}`

File: worlds/v6/Rules.py [hunk 4]
```diff
@@ -1,43 +1,44 @@
...
-    add_rule(world.get_location("NPC Trinket", player),
-             lambda state: state.can_reach("Laboratory", 'Region', player) or
-                           (state.can_reach("The Tower", 'Region', player) and
-                            state.can_reach("Space Station 2", 'Region', player) and
-                            state.can_reach("Warp Zone", 'Region', player)))
+    add_rule(multiworld.get_location("NPC Trinket", player), lambda state: can_reach_npc_trinket(state, player))
 
-    world.completion_condition[player] = lambda state: state.can_reach("V", 'Location', player)
+    # Victory condition
+    if multiworld.DoorCost[player] == 0:
```
Comment (N00byKing) [inline worlds/v6/Rules.py:41]: Please just use the original completion condition, no need to differentiate.

**Review by N00byKing — CHANGES_REQUESTED** [review]
Other than noted lgtm ^^

File: worlds/v6/__init__.py (hunk 3 above)
Comment (ThePhar) [reply worlds/v6/__init__.py:56]: Was a small optimization, but I can change it to that if you prefer.

Comment (N00byKing) [reply worlds/v6/__init__.py:56]: With a game as small as this i'd prefer it that way, tbh. Probably gets cached anyway

File: worlds/v6/Rules.py (hunk 4 above)
Comment (ThePhar) [reply worlds/v6/Rules.py:41]: Helps if my logic above was correct. Fixed.

File: worlds/v6/Rules.py (hunk 2 above)
Comment (ThePhar) [reply worlds/v6/Rules.py:27]: Fair enough, changed.

File: worlds/v6/Rules.py (hunk 1 above)
Comment (ThePhar) [reply worlds/v6/Rules.py:8]: Yeah that's a faulty brain moment. Thanks for the catch.

**Comment by ThePhar** [comment]
Guess, I should have been suspicious when it generated really fast lol. Made the changes and I'm running a test generation now. Will know once it's done.

**Comment by ThePhar** [comment]
Successful generation with 1000 players with the following YAML. Included spoiler.

```yaml
VVVVVV:
  progression_balancing: 0
  accessibility: locations
  MusicRandomizer: false
  AreaRandomizer: true
  DoorCost: random
  AreaCostRandomizer: true
  death_link: 'false'
  DeathLinkAmnesty: 15
description: 'Generated by https://archipelago.gg/'
game: VVVVVV
name: Player{PLAYER}
```

[AP_56811105121036757111_Spoiler.txt](https://github.com/ArchipelagoMW/Archipelago/files/10189284/AP_56811105121036757111_Spoiler.txt)

**Comment by ThePhar** [comment]
If you wanna take one final look, @N00byKing.

**Comment by N00byKing** [comment]
LGTM, thanks mate!

**Review by N00byKing — APPROVED** [review]

**Comment by ThePhar** [comment]
Noticed a major issue with the rules on this during another test. Going to move to Draft until I fix it.

**Pull request closed without merging**